util/git/creds.go

```diff
@@ -1,6 +1,88 @@
 package git
 
-// client credentials for Git HTTPS connection
-type Creds struct {
-	Username, Password string
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"strings"
+
+	log "github.com/sirupsen/logrus"
+)
+
+type Creds interface {
+	Environ() (io.Closer, []string, error)
+}
+
+// nop implementation
+type NopCloser struct {
+}
+
+func (c NopCloser) Close() error {
+	return nil
+}
+
+type NopCreds struct {
+}
+
+func (c NopCreds) Environ() (io.Closer, []string, error) {
+	return NopCloser{}, nil, nil
+}
+
+// HTTPS creds implementation
+type HTTPSCreds struct {
+	username, password string
+}
+
+func NewHTTPSCreds(username, password string) HTTPSCreds {
+	return HTTPSCreds{username, password}
+}
+
+func (c HTTPSCreds) Environ() (io.Closer, []string, error) {
+	return NopCloser{}, []string{
+		fmt.Sprintf("GIT_ASKPASS=%s", "git-ask-pass.sh"),
+		fmt.Sprintf("GIT_USERNAME=%s", c.username),
+		fmt.Sprintf("GIT_PASSWORD=%s", c.password),
+	}, nil
+}
+
+// SSH implementation
+type SSHCreds struct {
+	sshPrivateKey         string
+	insecureIgnoreHostKey bool
+}
+
+func NewSSHCreds(sshPrivateKey string, insecureIgnoreHostKey bool) SSHCreds {
+	return SSHCreds{sshPrivateKey, insecureIgnoreHostKey}
+}
+
+type sshPrivateKeyFile string
+
+func (f sshPrivateKeyFile) Close() error {
+	return os.Remove(string(f))
+}
+
+func (c SSHCreds) Environ() (io.Closer, []string, error) {
+	file, err := ioutil.TempFile("", "")
```

Comment:
Let's not use the default location of TempFile which will may be on-disk which is both slower and less secure (if clean up is not performed). For this type of interaction, it would be better to use shared memory. Let's copy what we do for running kubeconfigs, which is:

```golang
var (
	// location to use for generating temporary files, such as the kubeconfig needed by kubectl
	TempDir string
)

func init() {
	fileInfo, err := os.Stat("/dev/shm")
	if err == nil && fileInfo.IsDir() {
		TempDir = "/dev/shm"
	}
}
```

Reply:
Clever. I like it.

Reply:
Done